File: frontend/shader/contract/ts/solana.ts
import { JsonRpcResponse } from "../app/app";

// ============================================================================
// Solana Configuration
// ============================================================================

export interface SolanaConfig {
  rpcUrl: string;
  network: "mainnet" | "devnet" | "testnet" | "localnet";
}

// ============================================================================
// Solana API Response Types
// ============================================================================

export interface SolanaBalanceResponse {
  result?: {
    value?: number;
  };
}

/**
 * Solana Connector
 * Connects to Solana blockchain
 */
export class SolanaConnector {
  private readonly config: SolanaConfig;

  constructor(config: SolanaConfig) {
    this.config = config;
  }

  /**
   * Get balance for a Solana address
   * Uses Solana RPC API
   */
  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid Solana address");
    }

    try {
      const response = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 1,
          method: "getBalance",
          params: [address],
        }),
      });

      if (!response.ok) {
        throw new Error("Failed to fetch Solana balance");
      }

      const data = (await response.json()) as SolanaBalanceResponse & {
        error?: { message?: string };
      };
      if (data.error) {
        throw new Error(data.error.message || "RPC error");
      }

      // Return balance in lamports
      return data.result?.value?.toString() || "0";
    } catch (error) {
      throw new Error(
        `Solana getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  /**
   * Send SOL transaction
   * Note: This requires proper keypair signing
   */
  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string // in lamports
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress) || !this.isValidAddress(toAddress)) {
      throw new Error("Invalid Solana address");
    }

    const amountNum = Number.parseInt(amount, 10);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // Get recent blockhash
      const blockhashResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 1,
          method: "getLatestBlockhash",
          params: [{ commitment: "finalized" }],
        }),
      });

      if (!blockhashResponse.ok) {
        throw new Error("Failed to get recent blockhash");
      }

      const blockhashData =
        (await blockhashResponse.json()) as JsonRpcResponse & {
          result?: { value?: { blockhash?: string } };
        };
      if (blockhashData.error) {
        throw new Error(blockhashData.error.message || "RPC error");
      }

      // Note: In production, you would need to:
      // 1. Load the keypair from fromAddress
      // 2. Create and sign the transaction
      // 3. Send the signed transaction
      // This is a simplified version that shows the structure
      const transaction = {
        message: {
          accountKeys: [fromAddress, toAddress],
          instructions: [
            {
              programId: "11111111111111111111111111111111", // System Program
              accounts: [
                { pubkey: fromAddress, writable: true, signer: true },
                { pubkey: toAddress, writable: true, signer: false },
              ],
              data: Buffer.from([
                2,
                0,
                0,
                0, // Transfer instruction
                ...Buffer.from(
                  amountNum.toString(16).padStart(16, "0"),
                  "hex"
                ),
              ]).toString("base64"),
            },
          ],
          recentBlockhash: blockhashData.result?.value?.blockhash || "",
        },
      };

      // Send transaction (requires signing in production)
      const sendResponse = await fetch(this.config.rpcUrl, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          jsonrpc: "2.0",
          id: 2,
          method: "sendTransaction",
          params: [
            // In production, this would be a signed transaction
            JSON.stringify(transaction),
            { encoding: "base64", skipPreflight: false },
          ],
        }),
      });

      if (!sendResponse.ok) {
        throw new Error("Failed to send Solana transaction");
      }

      const sendData = (await sendResponse.json()) as JsonRpcResponse;
      if (sendData.error) {
        throw new Error(sendData.error.message || "Transaction failed");
      }

      return sendData.result as string;
    } catch (error) {
      throw new Error(
        `Solana sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  /**
   * Validate Solana address (base58, 32-44 chars)
   */
  isValidAddress(address: string): boolean {
    // Basic validation - Solana addresses are base58 encoded
    return address.length >= 32 && address.length <= 44;
  }

  /**
   * Bridge tokens from Solana to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "Solana";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["solana", "sol"];
  }
}

File: store/rust/rust/solana.rs
//! Solana connector for interacting with Solana blockchain

use crate::{BridgeError, BridgeResult, SolanaNetwork, solana_rpc_call};
use serde::{Deserialize, Serialize};

/// Solana connector for interacting with Solana blockchain
pub struct SolanaConnector {
    rpc_url: String,
    network: SolanaNetwork,
}

impl SolanaConnector {
    /// Create new Solana connector
    pub fn new(rpc_url: &str, network: SolanaNetwork) -> BridgeResult<Self> {
        Ok(Self { rpc_url: rpc_url.to_string(), network })
    }

    /// Get balance for a Solana address
    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        // Validate Solana address (base58, 32-44 chars)
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        let params = serde_json::json!([address, {"encoding": "jsonParsed"}]);
        let result = solana_rpc_call(&self.rpc_url, "getBalance", params)?;

        let balance = result
            .get("value")
            .and_then(|v| v.as_u64())
            .ok_or_else(|| BridgeError::DeserializationError("Invalid balance format".to_string()))?;

        Ok(balance)
    }

    /// Send SOL transaction
    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in lamports
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) {
            return Err(BridgeError::InvalidAddress(from_address.to_string()));
        }
        if !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress(to_address.to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // Note: Solana transactions require signing with a private key.
        // This implementation assumes the RPC endpoint has access to the wallet.
        // In production, you'd need to:
        // 1. Get recent blockhash
        // 2. Create transfer instruction
        // 3. Build and sign transaction
        // 4. Send transaction

        // For now, we'll use a simplified approach that requires the RPC to handle signing
        // This is typically done via wallet integration or by sending a pre-signed transaction
        // Here we return an error indicating that signing is required
        Err(BridgeError::TransactionFailed(
            "Solana transactions require private key signing. Use a wallet-integrated RPC or provide a signed transaction.".to_string()
        ))
    }

    /// Validate Solana address (base58)
    pub fn is_valid_address(&self, address: &str) -> bool {
        // Basic validation - Solana addresses are base58 encoded, 32-44 chars
        // In production, use proper base58 decoding
        address.len() >= 32 && address.len() <= 44
    }

    /// Get transaction info
    pub fn get_transaction(&self, signature: &str) -> BridgeResult<SolanaTransaction> {
        let params = serde_json::json!([signature, {"encoding": "jsonParsed", "maxSupportedTransactionVersion": 0}]);
        let result = solana_rpc_call(&self.rpc_url, "getTransaction", params)?;

        let slot = result.get("slot").and_then(|s| s.as_u64()).unwrap_or(0);

        let block_time = result.get("blockTime").and_then(|bt| bt.as_i64());

        // Extract amount and fee from transaction
        let amount = result
            .get("meta")
            .and_then(|m| m.get("preBalances"))
            .and_then(|pb| pb.as_array())
            .and_then(|arr| {
                if arr.len() >= 2 {
                    let pre = arr[0].as_u64()?;
                    let post = arr[1].as_u64()?;
                    Some(post.saturating_sub(pre))
                } else {
                    None
                }
            })
            .unwrap_or(0);

        let fee = result
            .get("meta")
            .and_then(|m| m.get("fee"))
            .and_then(|f| f.as_u64())
            .unwrap_or(0);

        Ok(SolanaTransaction {
            signature: signature.to_string(),
            slot,
            block_time,
            amount,
            fee,
        })
    }

    /// Get network
    pub fn get_network(&self) -> SolanaNetwork {
        self.network
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SolanaTransaction {
    pub signature: String,
    pub slot: u64,
    pub block_time: Option<i64>,
    pub amount: u64,
    pub fee: u64,
}

